include/albatross/src/models/patchwork_gp.hpp

```diff
@@ -0,0 +1,374 @@
+      CovFunc, PatchworkGaussianProcess<CovFunc, PatchworkFunctions>>;
+
+  PatchworkGaussianProcess() : Base(){};
+  PatchworkGaussianProcess(CovFunc &covariance_function)
+      : Base(covariance_function){};
+  PatchworkGaussianProcess(CovFunc &covariance_function,
+                           PatchworkFunctions patchwork_functions)
+      : Base(covariance_function), patchwork_functions_(patchwork_functions){};
+
+  template <typename X, typename Y>
+  Eigen::MatrixXd patchwork_covariance_matrix(const std::vector<X> &xs,
+                                              const std::vector<Y> &ys) const {
+    auto patchwork_caller = [&](const auto &x, const auto &y) {
+      return PatchworkCaller::call(this->covariance_function_, x, y);
+    };
+
+    return compute_covariance_matrix(patchwork_caller, xs, ys);
+  };
+
+  template <typename FitModelType, typename GroupKey>
+  auto
+  from_fit_models(const Grouped<GroupKey, FitModelType> &fit_models) const {
+
+    auto boundary_function = [&](const GroupKey &x, const GroupKey &y) {
+      return patchwork_functions_.boundary(x, y);
+    };
+
+    const auto boundary_features =
+        build_boundary_features(boundary_function, fit_models.keys());
+
+    using BoundaryFeatureType = typename std::decay<typename decltype(
+        boundary_features)::value_type>::type;
+    using PatchworkFitType =
+        Fit<PatchworkGPFit<FitModelType, GroupKey, BoundaryFeatureType>>;
+    using ReturnType = FitModel<PatchworkGaussianProcess, PatchworkFitType>;
+
+    if (fit_models.size() == 1) {
+      return ReturnType(*this, PatchworkFitType(fit_models));
+    }
+
+    assert(boundary_features.size() > 0);
+
+    // C_bb is the covariance matrix between all boundaries, it will
+    // have a lot of zeros so it could be decomposed more efficiently
+    const Eigen::MatrixXd C_bb =
+        patchwork_covariance_matrix(boundary_features, boundary_features);
+    const auto C_bb_ldlt = C_bb.ldlt();
+
+    // C_dd is the large block diagonal matrix, with one block for each model
+    // for which we already have an efficient way of computing the inverse.
+    auto get_train_covariance = [](const auto &fit_model) {
+      return fit_model.get_fit().train_covariance;
+    };
+    const auto C_dd = fit_models.apply(get_train_covariance);
+
+    // C_db holds the covariance between each model and all boundaries.
+    // The actual storage is effectively a map with values which correspond
+    // to the covariance between that model's features and the boundaries.
+    auto C_db_one_group = [&](const auto &key, const auto &fit_model) {
+      const auto group_features =
+          as_group_features(key, fit_model.get_fit().train_features);
+      return this->patchwork_covariance_matrix(group_features,
+                                               boundary_features);
+    };
+    const auto C_db = fit_models.apply(C_db_one_group);
+    const auto C_dd_inv_C_db = block_diag_solve(C_dd, C_db);
```

Comment:
I imagine this line and the following ones relate to certain equations in the paper. Can they be referenced, if so?